presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
     {
         Optional<ConnectorOutputMetadata> finishTable(Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics);
     }
+
+    public interface PartitionCommitter
+    {
+        void commitPartition(int partitionId, Collection<Slice> fragments);
+    }
+
+    private static class LifespanStageStateTracker
+    {
+        private final Map<LifespanStage, LifespanStageState> unrecoverableLifespanStageStates = new HashMap<>();
+        private final Map<LifespanStage, Map<Integer, LifespanStageState>> unfinishedRecoverableLifespanStageStates = new HashMap<>();
+        private final Map<LifespanStage, LifespanStageState> finishedRecoverableLifespanStages = new HashMap<>();
+
+        private final PartitionCommitter partitionCommitter;
+        private final JsonCodec<TableCommitContext> tableCommitContextCodec;
+
+        public LifespanStageStateTracker(PartitionCommitter partitionCommitter, JsonCodec<TableCommitContext> tableCommitContextCodec)
+        {
+            this.partitionCommitter = requireNonNull(partitionCommitter, "partitionCommitter is null");
+            this.tableCommitContextCodec = requireNonNull(tableCommitContextCodec, "tableCommitContextCodec is null");
+        }
+
+        public void update(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanStage lifespanStage = LifespanStage.fromTableCommitContext(tableCommitContext);
+            if (finishedRecoverableLifespanStages.containsKey(lifespanStage)) {
+                return;
+            }
+
+            switch (tableCommitContext.getCommitGranularity()) {
+                case TABLE:
+                    unrecoverableLifespanStageStates.putIfAbsent(lifespanStage, new LifespanStageState());
+                    unrecoverableLifespanStageStates.get(lifespanStage).update(page, false);
+                    break;
+                case PARTITION:
+                    unfinishedRecoverableLifespanStageStates.putIfAbsent(lifespanStage, new HashMap<>());
+                    Map<Integer, LifespanStageState> groupStageStatesPerTask = unfinishedRecoverableLifespanStageStates.get(lifespanStage);
```

Comment:
what's `GroupStage` ?

Reply:
This is a side effect of a rename. `LifespanStage` used to be named as `GroupStage` (for driver group). :D